crates/sui-core/src/authority/authority_store.rs

```diff
@@ -329,17 +332,30 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
     pub async fn get_transaction_envelope(
         &self,
         object_ref: &ObjectRef,
-    ) -> Result<Option<TransactionEnvelope<S>>, SuiError> {
+    ) -> SuiResult<Option<TransactionEnvelope<S>>> {
         let transaction_option = self
             .lock_service
             .get_lock(*object_ref)
             .await?
             .ok_or(SuiError::TransactionLockDoesNotExist)?;
 
+        // Returns None if either no TX with the lock, or TX present but no entry in transactions table.
+        // However we retry a couple times because the TX is written after the lock is acquired, so it might
+        // just be a race.
         match transaction_option {
             Some(tx_digest) => {
-                return Ok(self.transactions.get(&tx_digest)?);
-                // .expect("Stored a lock without storing transaction?"),
+                let retry_strategy = ExponentialBackoff::from_millis(20).map(jitter).take(3);
+                let mut tx_option = None;
+                for duration in retry_strategy {
+                    tx_option = self.transactions.get(&tx_digest)?;
+                    if tx_option.is_some() {
+                        break;
+                    }
+                    // Wait to retry
+                    tokio::time::sleep(duration).await;
```

Comment:
It seems like we are doing a wasted wait of 8s in the third attempt before bailing out?

Reply:
Hmmm that's true.  Let me see if we can get rid of this easily.
BTW it's not 8s... it would be 80ms

Reply:
I think it'll be 8000ms as the base of the exponential function is 20 here i.e. 20^x and delays are [20^1, 20^2, 20^3]. Also, checking the test here https://github.com/srijs/rust-tokio-retry/blob/master/src/strategy/exponential_backoff.rs 
We may want to use ExponentialBackoff::from_millis(2) .factor(10) to get [20, 40, 80]

Reply:
Ah ok thanks.  I'm about to push a change and I'll also change the default strategy.